transports/websocket-websys/src/web_context.rs

```
@@ -1,57 +1,185 @@
+    }
+
+    fn is_supported(&self) -> bool {
+        match self.0.load(Ordering::SeqCst) {
+            0x01 => true,
+            0x02 => false,
+            _ => {
+                let global = js_sys::global();
+                let is_supported = has(&global, "WebSocket");
+
+                let value = if is_supported { 0x01 } else { 0x02 };
+
+                self.0.store(value, Ordering::SeqCst);
+
+                is_supported
+            }
+        }
+    }
+}
+
 /// Web context that abstract the window vs web worker API
 #[derive(Debug)]
 pub(crate) enum WebContext {
     Window(web_sys::Window),
     Worker(web_sys::WorkerGlobalScope),
+    Unknown(Global),
 }
 
 impl WebContext {
     pub(crate) fn new() -> Option<Self> {
-        match window() {
-            Some(window) => Some(Self::Window(window)),
-            None => {
-                #[wasm_bindgen]
-                extern "C" {
-                    type Global;
-
-                    #[wasm_bindgen(method, getter, js_name = WorkerGlobalScope)]
-                    fn worker(this: &Global) -> JsValue;
-                }
-                let global: Global = js_sys::global().unchecked_into();
-                if !global.worker().is_undefined() {
-                    Some(Self::Worker(global.unchecked_into()))
-                } else {
-                    None
-                }
-            }
+        if let Some(window) = window() {
+            return Some(Self::Window(window));
         }
+
+        let global: Global = js_sys::global().unchecked_into();
+
+        if has(&global, "WorkerGlobalScope") {
+            return Some(Self::Worker(global.worker().unchecked_into()));
+        }
+
+        // check if the `setInterval` and `clearInterval` are available globally
+        if has(&global, "setInterval") && has(&global, "clearInterval") {
+            return Some(Self::Unknown(global));
+        }
+
+        None
+    }
+
+    pub(crate) fn is_websocket_supported() -> bool {
```

Why do we need this? Wouldn't it be better to make `new` more performant by caching if it is not supported?

---

Yes and no, I see how this has utility in `new` (an oversight of me not adding it). Still, the explicit call to this is to improve user ergonomics. `dial()` will panic instead of silently continuing with the error that it is an invalid address.

The connection is only formed once a successful `WebSocket` has been established, so the user won't know what to do if it isn't —this way, they will.

---

I don't think we should panic in `dial`. If we can't dial because the JS environment doesn't support it, we should return an error instead. Whether or not to panic on that should be left to the user code.

---

We should be able to also achieve a good UX by encapsulating creation of `Connection` more such that we can return an error if we can't bind to the required JS functions.